### PR TITLE
Leaderboard Use Case Implementation

### LINKED ISSUE #19: Write JUnit Tests for Leaderboard Use Case
- [ ] Write tests to cover MongoDB access code
- [ ] Write tests to cover interactor functionality
- [ ] Write tests to cover interface adapter functionality
- [ ] Write tests to ensure the whole use case works.

### LINKED ISSUE #23: Leaderboard button not working correctly
Ensure the button click does the following things:
- [ ] Leaderboard button should switch to new view
- [ ] Leaderboard button should get the most up to date leaderboard information from the database
- [ ] Leaderboard button should have a nice error message if the use case fails somehow (i.e. the database can't be accessed)

### PR DESCRIPTION
**Pull request by YoshikuYuu**
Implement the leaderboard use case and edit Profile entity and UserFileDAO appropriately.

### CODE REVIEW

**Review by connorsicheri — APPROVED** [review]
This is a great contribution and use case overall. Your coding conventions are very solid, and make you code easy to follow. 

From running the code on my machine, it seems to be working correctly, aside from the database not yet being on the cloud. 
One issue is that modification to the profile dataAccessInterface has caused the tests made for the FileUserDAO and Profile Interactor to fail. This can be modified later on

The limited tests available are great, but a more inclusive test profile is required.

The clean architecture is strictly adhered too, which will make this easy to write future tests for, and to integrate to the rest of our project.

The SOLID principles seem to be used throughout your code, for example having all of your higher level classes rely on interfaces, instead of concrete implementations.

Your javadoc seems to be limited, so adding more would be a great improvement

**Review by lifers — COMMENTED** [review]
Overall good naming conventions as shown in the course, and no unnecessary access opened.

I tried running the code on my computer, most of the things worked except the online database connection. I think this can be done in another issue. Also there is an import missing in QuestionUseCaseFactory, please fix before merging.

Note that there is a breaking change in ProfileDataAccessInterface, so tests have to be rewritten in another issue.

Speaking of tests, it mostly covers the MongoDBDataAccessObject and shown some coverage for the leaderboard use case interactors.

Clean architecture and SOLID principle is nicely upheld, making the test code simpler and easier to maintain. There is a notable lack of Javadoc, however this can be added in another issue. Overall, this is a big PR that implements a very complicated feature, good job to the writer and I think it's ready to merge after we fixed the import problem.

File: src/app/QuestionUseCaseFactory.java [hunk 1]
```diff
@@ -2,12 +2,12 @@
 
 import data_access.FileUserDataAccessObject;
 import interface_adapter.answer_question.AnswerQuestionController;
-import interface_adapter.answer_question.AnswerQuestionPresenter;
```
Comment (lifers) [inline src/app/QuestionUseCaseFactory.java:5]: Don't delete this

**Review by lifers — CHANGES_REQUESTED** [review]

**Review by lifers — APPROVED** [review]
Alright all good.

**Pull request merged**